lib/gooddata/models/data_source.rb

```diff
@@ -0,0 +1,648 @@
+
+require_relative '../rest/resource'
+
+module GoodData
+  class DataSource < Rest::Resource
+    attr_accessor :connection_info
+
+    DATA_SOURCES_URL = '/gdc/dataload/dataSources'
+    SNOWFLAKE = 'snowflake'
+    REDSHIFT = 'redshift'
+    BIGQUERY = 'bigQuery'
+    GENERIC = 'generic'
+    S3 = 's3'
+    ADS = 'ads'
+
+    class << self
+      # Get all data sources or get a specify data source from data source identify
+      # Expected parameter value:
+      # - :all return all data sources
+      # - :data_source_id return a data source with the specify data source identify
+      def [](id = :all, options = { client: GoodData.client })
+        c = GoodData.get_client(options)
+
+        if id == :all
+          data = c.get(DATA_SOURCES_URL)
+          data['dataSources']['items'].map do |ds_data|
+            c.create(DataSource, ds_data)
+          end
+        else
+          c.create(DataSource, c.get("#{DATA_SOURCES_URL}/#{id}"))
+        end
+      end
+
+      # Get a specify data source from data source identify
+      def from_id(id, options = { client: GoodData.client })
+        DataSource[id, options]
+      end
+
+      # Get a specify data source from data source alias
+      def from_alias(data_source_alias, options = { client: GoodData.client })
+        data_sources = all(options)
+        data_sources.find do |data_source|
+          data_source.alias == data_source_alias
+        end
+      end
+
+      # Get all data sources
+      def all(options = { client: GoodData.client })
+        DataSource[:all, options]
+      end
+
+      # Create data source from json
+      # Expected keys:
+      # - :name (mandatory)
+      # - :alias (optional)
+      # - :prefix (optional)
+      # - :connectionInfo (mandatory)
+      # - :client (mandatory)
+      def create(opts)
+        ds_name = opts[:name]
+        ds_alias = opts[:alias]
+        ds_prefix = opts[:prefix]
+        ds_connection_info = opts[:connectionInfo]
+
+        GoodData.logger.info "Creating data source '#{ds_name}'"
+        fail ArgumentError, 'Data source name has to be provided' if ds_name.nil? || ds_name.blank?
+        fail ArgumentError, 'Data source connection info has to be provided' if ds_connection_info.nil?
+        fail ArgumentError, 'Data source does not support' unless support_data_source(ds_connection_info)
+
+        json = {
+          'dataSource' => {
+            'name' => ds_name,
+            'connectionInfo' => ds_connection_info
+          }
+        }
+        json['dataSource']['alias'] = ds_alias if ds_alias
+        json['dataSource']['prefix'] = ds_prefix if ds_prefix
+
+        # Create data source
+        c = GoodData.get_client(opts)
+        res = c.post(DATA_SOURCES_URL, json)
+
+        # create the public facing object
+        c.create(DataSource, res)
+      end
+
+      # Check whether the data source connection is supported
+      def support_data_source(connection_info_json)
```

Comment:
Is this functions necessary? When backend supports new data source type, we will update SDK? I think we should rely on the validation on server side.

Reply:
Remove the function. We will rely on the validation on server side.